packages/devtools_app/lib/src/service/editor/editor_client.dart

```diff
@@ -57,3 +78,208 @@ abstract class EditorClient {
+        _supportsHotRestart = isRegistered;
+      } else if (method == EditorMethod.openDevToolsPage.name) {
+        _supportsOpenDevToolsPage = isRegistered;
+        _supportsOpenDevToolsForceExternal =
+            capabilities?['supportsForceExternal'] == true;
+      } else {
+        return;
+      }
+      _editorServiceChangedController.add(method);
+    });
+
+    final editorKindMap = EditorEventKind.values.asNameMap();
+    _dtd.onEvent(editorStreamName).listen((data) {
+      final kind = editorKindMap[data.kind];
+      switch (kind) {
+        case null:
+          // Unknown event. Use null here so we get exhaustiveness checking for
+          // the rest.
+          break;
+        case EditorEventKind.deviceAdded:
+          _eventController.add(DeviceAddedEvent.fromJson(data.data));
+        case EditorEventKind.deviceRemoved:
+          _eventController.add(DeviceRemovedEvent.fromJson(data.data));
+        case EditorEventKind.deviceChanged:
+          _eventController.add(DeviceChangedEvent.fromJson(data.data));
+        case EditorEventKind.deviceSelected:
+          _eventController.add(DeviceSelectedEvent.fromJson(data.data));
+        case EditorEventKind.debugSessionStarted:
+          _eventController.add(DebugSessionStartedEvent.fromJson(data.data));
+        case EditorEventKind.debugSessionChanged:
+          _eventController.add(DebugSessionChangedEvent.fromJson(data.data));
+        case EditorEventKind.debugSessionStopped:
+          _eventController.add(DebugSessionStoppedEvent.fromJson(data.data));
+      }
+    });
+    await Future.wait([
+      _dtd.streamListen('Service'),
+      _dtd.streamListen(editorServiceName),
+    ]);
+  }
+
+  /// Close the connection to DTD.
+  @override
+  Future<void> close() => _dtd.close();
+
+  @override
+  bool get supportsGetDevices => _supportsGetDevices;
+  @override
+  bool get supportsGetDebugSessions => _supportsGetDebugSessions;
+  @override
+  bool get supportsSelectDevice => _supportsSelectDevice;
+  @override
+  bool get supportsHotReload => _supportsHotReload;
+  @override
+  bool get supportsHotRestart => _supportsHotRestart;
+  @override
+  bool get supportsOpenDevToolsPage => _supportsOpenDevToolsPage;
+  @override
+  bool get supportsOpenDevToolsForceExternal =>
+      _supportsOpenDevToolsForceExternal;
+
+  var _supportsGetDevices = false;
+  var _supportsGetDebugSessions = false;
+  var _supportsSelectDevice = false;
+  var _supportsHotReload = false;
+  var _supportsHotRestart = false;
+  var _supportsOpenDevToolsPage = false;
+  var _supportsOpenDevToolsForceExternal = false;
+
+  /// A stream of [EditorEvent]s from the editor.
+  @override
+  Stream<EditorEvent> get event => _eventController.stream;
+  final _eventController = StreamController<EditorEvent>();
+
+  /// A stream of events of when editor services are registrered or
+  /// unregistered.
+  ///
+  /// The values represent the method names that were registered or
+  /// unregistered.
```

Comment:
I thought we don't, since we really just care when it changed and then we can read the boolean flags (perhaps once the postMessage version is done we could use the capabilities directly from the event and toggle state in the widget). However, you made me look at this again and I saw that we would treat the unregistration of `editor.GetDevices` as if it was a registration and then try to call it. Hopefully that never happens (the editor shouldn't be disconnecting from DTD while the sidebar is visible!), but I've updated this in 83fd3e7df to be more structured (the stream is one of two subclasses for registration/unregistration with additional fields).